### PR TITLE
 XP Tracker: Add skill-appear order options

### PR DESCRIPTION
**Pull request by angus-beef**
This will allow the user to select from three display order options. 
These options give more control to the user and will let them determine the order in which skills are displayed in the skill pane.

1. Least recent - current behaviour
2. Most recent - most recent skill to gain exp goes to the top.
3. Alphabetically - in case there is a preference for this.

### CODE REVIEW

**Comment by Lalonde33** [comment]
BPGD I should create a feature req but anyway you could add an option to auto expand skills on xp tracker. Annoying to start doing something and have to stop to go open it up. Thanks.

**Comment by BFGD** [comment]
That should be pretty straight forward. I'll make another pr for it sometime unless someone else gets around to it first.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 1]
```diff
@@ -185,6 +186,26 @@ public void mouseReleased(MouseEvent e)
 		add(container, BorderLayout.CENTER);
 	}
 
+	private static BufferedImage createHoverImage(BufferedImage image)
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Nothing else you changed uses this function, and I don't see a difference. Why did you move this?

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBoxOrderState.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		{
+			return;
+		}
+
+		SkillOrderType skillOrderType = config.reorder();
+
+		switch (skillOrderType)
+		{
+			case LEAST_RECENT:
+				infoBoxes.sort(Comparator.comparing(o -> o.addedOrder));
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBoxOrderState.java]: Instead of o -> o.addedOrder, you can do ClassName::methodName. This makes it clearer as to where this function comes from.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBoxOrderState.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+			case ALPHABETICAL:
+				infoBoxes.sort(Comparator.comparing(o -> o.skill.getName()));
+				break;
+		}
+	}
+
+	void setInfoBoxOrderState(XpInfoBox infoBox, Component[] oldBoxOrder)
+	{
+		LinkedList<XpInfoBox> boxes = new LinkedList<>();
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBoxOrderState.java]: Prefer interface on type declaration

`List<XpInfoBox> boxes = new LinkedList()`

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java [hunk 4]
```diff
@@ -51,11 +52,19 @@
 	private final Map<Skill, XpInfoBox> infoBoxes = new HashMap<>();
 	private final JLabel totalXpGained = new JLabel();
 	private final JLabel totalXpHr = new JLabel();
+	private final JPanel infoBoxPanel = new JPanel();
+	private XpTrackerConfig config;
+	private LinkedList<XpInfoBox> cachedBoxOrder;
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java]: Prefer interface `List<XpInfoBox>`

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java (hunk 1 above)
Comment (BFGD) [reply runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: I believe it was due to code styling preferences automatically hoisting static methods to the top of the file. I left it in because I thought it made sense. I could revert the change.

Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: That explains all the sudden whitespace removals in the xp tracker plugin around the private variables.

**Comment by deathbeam** [comment]
I don't think this should be added before @psikoi 's change is fully reviewed. Actually, all UI changes should be deprioritized and his should be prioritized, because most of them is irellevant anyway when/if his changes will go in.

**Comment by BFGD** [comment]
That's fair. I'll keep an eye out and update the PR when #1419 is merged.
